### PR TITLE
Fix double free when building Gtest/GMock in shared libraries and lin…

### PR DESCRIPTION
**Pull request by Romain-Geissler**
…king a test executable with both.

Hi,

This fixes issue #930. It is based on what was done in PR #1286 which itself was based on PR #1059 .

The only change from #1286 is that in the tests, the shared library gmock_shared_main is not modified at all, so that the build can go on, and tests are not broken. Tested locally, it fixes the core dump people have when linking both gmock and gtest using shared libraries.

Cheers,
Romain

### CODE REVIEW

**Comment by Romain-Geissler** [comment]
Note: the build failed on 2 OS X hosts in the Travis jobs, but this is certainly not due to this pull request. If fails the same way on the last commit of branch master.

File: googlemock/CMakeLists.txt [hunk 1]
```diff
@@ -86,16 +86,23 @@ endif()
...
-            "${cxx_strict}"
-            "${gtest_dir}/src/gtest-all.cc"
-            src/gmock-all.cc)
-
-cxx_library(gmock_main
-            "${cxx_strict}"
-            "${gtest_dir}/src/gtest-all.cc"
-            src/gmock-all.cc
-            src/gmock_main.cc)
+if (MSVC)
```
Comment (agirault) [inline googlemock/CMakeLists.txt]: Whys is it necessary to keep doing it that way for MSVC? Wouldn't the `target_link_libraries` way work fine?

Comment (Romain-Geissler) [reply googlemock/CMakeLists.txt]: Apparently not if I believe #1286. But the original build issue seems not available anymore. I have honestly not tried myself (and don't have any Windows machine beyond the continuous integration ones).

**Comment by gennadiycivil** [comment]
@Romain-Geissler Please 
Thank you very much for this contribution. Please provide testing output that clearly demonstrates before/after.

File: googlemock/CMakeLists.txt [hunk 2]
```diff
@@ -175,23 +182,33 @@ if (gmock_build_tests)
...
-    cxx_test_with_flags(gmock_use_own_tuple_test "${cxx_use_own_tuple}"
-      gmock_main_use_own_tuple test/gmock-spec-builders_test.cc)
+    if (MSVC_VERSION LESS 1600)  # 1600 is Visual Studio 2010.
+      # Visual Studio 2010, 2012, and 2013 define symbols in std::tr1 that
+      # conflict with our own definitions. Therefore using our own tuple does not
+      # work on those compilers.
+      cxx_library(gmock_main_use_own_tuple "${cxx_use_own_tuple}"
+        "${gtest_dir}/src/gtest-all.cc" src/gmock-all.cc src/gmock_main.cc)
+
+      cxx_test_with_flags(gmock_use_own_tuple_test "${cxx_use_own_tuple}"
```
Comment (Arfrever) [inline googlemock/CMakeLists.txt]: Previously `gmock_use_own_tuple_test` was also built with non-MSVC compiler.
I have tested changed version of this patch with `gmock_use_own_tuple_test` building enabled on GNU/Linux with GCC and `gmock_use_own_tuple_test` test passed.
I think that these 2 lines should be copied to non-MSVC branch below.

**Comment by Arfrever** [comment]
This patch fixes tests of Protobuf, which otherwise fail with segmentation fault and abort.
Please merge (some version of) this patch.

**Comment by gennadiycivil** [comment]
Thank you again for this. Could you please provide testing evidence to prove that this in fact "This patch fixes tests of Protobuf, which otherwise fail with segmentation fault and abort"

**Comment by Arfrever** [comment]
This was reproduced by both Toralf Förster and me:
https://bugs.gentoo.org/631698
https://bugs.gentoo.org/644044 (Duplicate of 631698)

From Toralf Förster's log in the latter bug:
```
make  check-TESTS
make[3]: Entering directory '/var/tmp/portage/dev-libs/protobuf-3.5.1.1/work/protobuf-3.5.1.1-abi_x86_64.amd64/src'
make[4]: Entering directory '/var/tmp/portage/dev-libs/protobuf-3.5.1.1/work/protobuf-3.5.1.1-abi_x86_64.amd64/src'
/var/tmp/portage/dev-libs/protobuf-3.5.1.1/work/protobuf-3.5.1.1/test-driver: line 107:  3393 Segmentation fault      "$@" > $log_file 2>&1
FAIL: protobuf-test
/var/tmp/portage/dev-libs/protobuf-3.5.1.1/work/protobuf-3.5.1.1/test-driver: line 107: 14114 Aborted                 "$@" > $log_file 2>&1
FAIL: protobuf-lazy-descriptor-test
/var/tmp/portage/dev-libs/protobuf-3.5.1.1/work/protobuf-3.5.1.1/test-driver: line 107: 14217 Aborted                 "$@" > $log_file 2>&1
FAIL: protobuf-lite-test
PASS: google/protobuf/compiler/zip_output_unittest.sh
PASS: google/protobuf/io/gzip_stream_unittest.sh
PASS: protobuf-lite-arena-test
============================================================================
Testsuite summary for Protocol Buffers 3.5.1
============================================================================
# TOTAL: 6
# PASS:  3
# SKIP:  0
# XFAIL: 0
# FAIL:  3
# XPASS: 0
# ERROR: 0
============================================================================
See src/test-suite.log
Please report to protobuf@googlegroups.com
============================================================================
make[4]: *** [Makefile:6758: test-suite.log] Error 1
make[4]: Leaving directory '/var/tmp/portage/dev-libs/protobuf-3.5.1.1/work/protobuf-3.5.1.1-abi_x86_64.amd64/src'
make[3]: *** [Makefile:6866: check-TESTS] Error 2
make[3]: Leaving directory '/var/tmp/portage/dev-libs/protobuf-3.5.1.1/work/protobuf-3.5.1.1-abi_x86_64.amd64/src'
make[2]: *** [Makefile:6972: check-am] Error 2
make[2]: Leaving directory '/var/tmp/portage/dev-libs/protobuf-3.5.1.1/work/protobuf-3.5.1.1-abi_x86_64.amd64/src'
make[1]: *** [Makefile:6974: check] Error 2
make[1]: Leaving directory '/var/tmp/portage/dev-libs/protobuf-3.5.1.1/work/protobuf-3.5.1.1-abi_x86_64.amd64/src'
make: *** [Makefile:1554: check-recursive] Error 1
```
(I got the same results.)

**Pull request merged**